In [1]:
# Add the dependencies.
import pandas as pd
import os

In [2]:
# Files to load
school_data_to_load = os.path.join("Resources", "schools_complete.csv")
student_data_to_load = os.path.join("Resources", "students_complete.csv")

### Read the school data file and store it in a Pandas DataFrame.

In [3]:
school_data_df = pd.read_csv(school_data_to_load)

### Read the school data file and store it in a Pandas DataFrame.

In [4]:
student_data_df = pd.read_csv(student_data_to_load)

In [5]:
student_data_df.tail()

Student ID     student_name gender grade         school_name  \
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  
39165             99          90  
39166             95          70  
39167             73          84  
39168             99          90  
39169             95          75

In [6]:
school_data_df.count()

School ID      15
school_name    15
type           15
size           15
budget         15
dtype: int64

In [7]:
student_data_df.count()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

In [8]:
school_data_df.isnull()

School ID  school_name   type   size  budget
0       False        False  False  False   False
1       False        False  False  False   False
2       False        False  False  False   False
3       False        False  False  False   False
4       False        False  False  False   False
5       False        False  False  False   False
6       False        False  False  False   False
7       False        False  False  False   False
8       False        False  False  False   False
9       False        False  False  False   False
10      False        False  False  False   False
11      False        False  False  False   False
12      False        False  False  False   False
13      False        False  False  False   False
14      False        False  False  False   False

In [9]:
# Determine if there are any missing values in the student data.
student_data_df.isnull()

Student ID  student_name  gender  grade  school_name  reading_score  \
0           False         False   False  False        False          False   
1           False         False   False  False        False          False   
2           False         False   False  False        False          False   
3           False         False   False  False        False          False   
4           False         False   False  False        False          False   
...           ...           ...     ...    ...          ...            ...   
39165       False         False   False  False        False          False   
39166       False         False   False  False        False          False   
39167       False         False   False  False        False          False   
39168       False         False   False  False        False          False   
39169       False         False   False  False        False          False   

       math_score  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
39165       False  
39166       False  
39167       False  
39168       False  
39169       False  

[39170 rows x 7 columns]

In [10]:
# Determine if there are not any missing values in the school data.
school_data_df.notnull()

School ID  school_name  type  size  budget
0        True         True  True  True    True
1        True         True  True  True    True
2        True         True  True  True    True
3        True         True  True  True    True
4        True         True  True  True    True
5        True         True  True  True    True
6        True         True  True  True    True
7        True         True  True  True    True
8        True         True  True  True    True
9        True         True  True  True    True
10       True         True  True  True    True
11       True         True  True  True    True
12       True         True  True  True    True
13       True         True  True  True    True
14       True         True  True  True    True

In [11]:
# Determine if there are not any missing values in the student data.
student_data_df.notnull().sum()

Student ID       39170
student_name     39170
gender           39170
grade            39170
school_name      39170
reading_score    39170
math_score       39170
dtype: int64

### Determine data types for the school DataFrame.

In [12]:
school_data_df.dtypes

School ID       int64
school_name    object
type           object
size            int64
budget          int64
dtype: object

### Determine data types for the student DataFrame.

In [13]:
student_data_df.dtypes

Student ID        int64
student_name     object
gender           object
grade            object
school_name      object
reading_score     int64
math_score        int64
dtype: object

In [14]:
# Add each prefix and suffix to remove to a list.
prefixes_suffixes = ["Dr. ", "Mr. ","Ms. ", "Mrs. ", "Miss ", " MD", " DDS", " DVM", " PhD"]

# Iterate through the "prefixes_suffixes" list and replace them with an empty space, "" when it appears in the student's name.
for word in prefixes_suffixes:
    student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")

<ipython-input-14-f7d7a5cf041f>:6: FutureWarning: The default value of regex will change from True to False in a future version.
  student_data_df["student_name"] = student_data_df["student_name"].str.replace(word,"")


### Combine the data into a single dataset.

In [15]:
school_data_complete_df = pd.merge(student_data_df, school_data_df, on=["school_name", "school_name"])

### Get the number of students and schools

In [16]:
student_count = school_data_complete_df.count()

# or 
student_count = school_data_complete_df['Student ID'].count()

In [17]:
school_count = school_data_df['School ID'].count()

#### From the merged dataFrame

In [18]:
school_count_2 = school_data_complete_df["school_name"].unique()
len(school_count_2)

15

### Get the SUM of a column

In [19]:
total_budget = school_data_df["budget"].sum()
total_budget

24649428

### Get AVG score

In [20]:
average_reading_score = school_data_complete_df['reading_score'].mean()

In [21]:
average_math_score = school_data_complete_df['math_score'].mean()

### Determine the Passing Grade

In [22]:
passing_math = school_data_complete_df["math_score"] >= 70
passing_reading = school_data_complete_df["reading_score"] >= 70

In [23]:
passing_math = school_data_complete_df[school_data_complete_df["math_score"] >= 70]
passing_math_count = passing_math['Student ID'].count()

In [24]:
passing_reading = school_data_complete_df[school_data_complete_df["reading_score"] >= 70]
passing_reading_count = passing_reading['Student ID'].count()

### Calculate percentage of students of passed Math and Reading

In [25]:
# Calculate the percent that passed math.
passing_math_percentage = passing_math_count / float(student_count) * 100
passing_reading_percentage = passing_reading_count / float(student_count) * 100

In [26]:
student_data_df

Student ID     student_name gender grade         school_name  \
0               0     Paul Bradley      M   9th   Huang High School   
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
3               3    Richard Scott      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  
0                 66          79  
1                 94          61  
2                 90          60  
3                 67          58  
4                 97          84  
...              ...         ...  
39165             99          90  
39166             95          70  
39167             73          84  
39168             99          90  
39169             95          75  

[39170 rows x 7 columns]

### Calculate the students who passed both math and reading.

In [27]:
passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]

In [28]:
overall_passing_math_reading_count = passing_math_reading.student_name.count()

### Calculate the overall passing percentage.

In [29]:
overall_passing_percentage = overall_passing_math_reading_count / student_count * 100
overall_passing_percentage

65.17232575950983

## New DataFrame district_summary_df

In [30]:
district_summary_df = pd.DataFrame([{
    "Total Schools": school_count,
    "Total Students": student_count,
    "Total Budget": total_budget,
    "Average Math Score": average_math_score,
    "Average Reading Score": average_reading_score,
    "% Passing Math": passing_math_percentage,
    "% Passing Reading": passing_reading_percentage,
    "% Overall Passing": overall_passing_percentage}])
district_summary_df

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

#### Define a function that calculates the percentage of students that passed both
#### math and reading and prints the passing percentage to the output when the function is called.

In [31]:
def passing_math_percent(pass_math_count, student_count):
    return pass_math_count / float(student_count) * 100

In [32]:
passing_math_count = 29370
total_student_count = 39170
passing_math_percent(passing_math_count, total_student_count)

74.9808526933878

#### Excercise not part of project

In [33]:
# A list of my grades.
my_grades = ['B', 'C', 'B' , 'D']
# Convert the my_grades to a Series
my_grades = pd.Series(my_grades)

In [34]:
# Change the grades by one letter grade.
my_grades.map({'B': 'A', 'C': 'B', 'D': 'C'})

0    A
1    B
2    A
3    C
dtype: object

### Formatting

In [35]:
# Using the format() function.
my_grades = [92.34, 84.56, 86.78, 98.32]

for grade in my_grades:
    print("{:.1f}".format(grade))

92.3
84.6
86.8
98.3


In [36]:
# Convert the numerical grades to a Series.
my_grades = pd.Series([92.34, 84.56, 86.78, 78.32])
my_grades
my_grades.map("{:.0f}".format)

0    92
1    85
2    87
3    78
dtype: object

In [37]:
student_data_df

Student ID     student_name gender grade         school_name  \
0               0     Paul Bradley      M   9th   Huang High School   
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
3               3    Richard Scott      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  
0                 66          79  
1                 94          61  
2                 90          60  
3                 67          58  
4                 97          84  
...              ...         ...  
39165             99          90  
39166             95          70  
39167             73          84  
39168             99          90  
39169             95          75  

[39170 rows x 7 columns]

### Format the "Total Students" to have the comma for a thousands separator.

In [38]:
district_summary_df["Total Students"] = district_summary_df["Total Students"].map("{:,}".format)

district_summary_df["Total Students"]

0    39,170
Name: Total Students, dtype: object

In [39]:
district_summary_df["Total Budget"] = district_summary_df["Total Budget"].map("{:,}".format)

district_summary_df["Total Budget"]

0    24,649,428
Name: Total Budget, dtype: object

In [40]:
# Format the columns.
district_summary_df["Average Math Score"] = district_summary_df["Average Math Score"].map("{:.1f}".format)

district_summary_df["Average Reading Score"] = district_summary_df["Average Reading Score"].map("{:.1f}".format)

district_summary_df["% Passing Math"] = district_summary_df["% Passing Math"].map("{:.0f}".format)

district_summary_df["% Passing Reading"] = district_summary_df["% Passing Reading"].map("{:.0f}".format)

district_summary_df["% Overall Passing"] = district_summary_df["% Overall Passing"].map("{:.0f}".format)


In [41]:
district_summary_df

Total Schools Total Students Total Budget Average Math Score  \
0             15         39,170   24,649,428               79.0   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                  81.9             75                86                65

### Reorder the columns in the order you want them to appear

In [42]:
# Same colums listed in the order we want
new_column_order = ["Total Schools", "Total Students", "Total Budget","Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

# Assign district summary df the new column order.
district_summary_df = district_summary_df[new_column_order]

### 4.8.1

**set school_name as the index and include column type**

Columns "school_name" and "type" are series inside dataframe "school_data_df"

Use series "school_name" as the index with ".set_index" and series "type" as the first column

In [43]:
per_school_types = school_data_df.set_index(["school_name"])["type"]

### Convert series to DataFrame

In [44]:
df = pd.DataFrame(per_school_types)

### Checking if school_data_df has the school_name as the index so we can reference the "size"

In [45]:
per_school_counts = school_data_df.set_index(['school_name'])['size']

### Checking if school_data_complete_df has the school_name as the index so we can reference the "size"

In [46]:
per_school_complete_counts = school_data_complete_df['school_name'].value_counts()

In [47]:
per_school_budget = school_data_df.set_index(['school_name'])['budget']

**Since both series have "school_name" as the index and both data types are int64**

In [48]:
per_school_capita = per_school_budget / per_school_counts

In [49]:
student_school_math = student_data_df.set_index(["school_name"])["math_score"]

In [50]:
per_school_averages = school_data_complete_df.groupby(["school_name"]).mean()

In [51]:
per_school_math = school_data_complete_df.groupby(["school_name"]).mean()["math_score"]
per_school_math

school_name
Bailey High School       77.048432
Cabrera High School      83.061895
Figueroa High School     76.711767
Ford High School         77.102592
Griffin High School      83.351499
Hernandez High School    77.289752
Holden High School       83.803279
Huang High School        76.629414
Johnson High School      77.072464
Pena High School         83.839917
Rodriguez High School    76.842711
Shelton High School      83.359455
Thomas High School       83.418349
Wilson High School       83.274201
Wright High School       83.682222
Name: math_score, dtype: float64

In [52]:
per_school_reading = school_data_complete_df.groupby(["school_name"]).mean()["reading_score"]
per_school_reading

school_name
Bailey High School       81.033963
Cabrera High School      83.975780
Figueroa High School     81.158020
Ford High School         80.746258
Griffin High School      83.816757
Hernandez High School    80.934412
Holden High School       83.814988
Huang High School        81.182722
Johnson High School      80.966394
Pena High School         84.044699
Rodriguez High School    80.744686
Shelton High School      83.725724
Thomas High School       83.848930
Wilson High School       83.989488
Wright High School       83.955000
Name: reading_score, dtype: float64

In [53]:
# Calculate the passing scores by creating a filtered DataFrame.
per_school_passing_math = school_data_complete_df[(school_data_complete_df["math_score"] >= 70)]
per_school_passing_math

Student ID     student_name gender grade         school_name  \
0               0     Paul Bradley      M   9th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
8               8     Michael Roth      M  10th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
8                 95          87          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[29370 rows x 11 columns]

In [54]:
# Calculate the passing scores by creating a filtered DataFrame.
per_school_passing_reading = school_data_complete_df[(school_data_complete_df["reading_score"] >= 70)]
per_school_passing_reading

Student ID     student_name gender grade         school_name  \
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[33610 rows x 11 columns]

In [55]:
per_school_passing_math = per_school_passing_math.groupby(["school_name"]).count()["student_name"]
per_school_passing_math

school_name
Bailey High School       3318
Cabrera High School      1749
Figueroa High School     1946
Ford High School         1871
Griffin High School      1371
Hernandez High School    3094
Holden High School        395
Huang High School        1916
Johnson High School      3145
Pena High School          910
Rodriguez High School    2654
Shelton High School      1653
Thomas High School       1525
Wilson High School       2143
Wright High School       1680
Name: student_name, dtype: int64

In [56]:
per_school_passing_reading = per_school_passing_reading.groupby(["school_name"]).count()["student_name"]

In [57]:
# Calculate the percentage of passing math and reading scores per school.
per_school_passing_math = per_school_passing_math / per_school_counts * 100
per_school_passing_math

school_name
Bailey High School       66.680064
Cabrera High School      94.133477
Figueroa High School     65.988471
Ford High School         68.309602
Griffin High School      93.392371
Hernandez High School    66.752967
Holden High School       92.505855
Huang High School        65.683922
Johnson High School      66.057551
Pena High School         94.594595
Rodriguez High School    66.366592
Shelton High School      93.867121
Thomas High School       93.272171
Wilson High School       93.867718
Wright High School       93.333333
dtype: float64

In [58]:
per_school_passing_reading = per_school_passing_reading / per_school_counts * 100
per_school_passing_reading

school_name
Bailey High School       81.933280
Cabrera High School      97.039828
Figueroa High School     80.739234
Ford High School         79.299014
Griffin High School      97.138965
Hernandez High School    80.862999
Holden High School       96.252927
Huang High School        81.316421
Johnson High School      81.222432
Pena High School         95.945946
Rodriguez High School    80.220055
Shelton High School      95.854628
Thomas High School       97.308869
Wilson High School       96.539641
Wright High School       96.611111
dtype: float64

In [59]:
# passed both math and reading.
per_passing_math_reading = school_data_complete_df[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70)]

per_passing_math_reading

Student ID     student_name gender grade         school_name  \
4               4       Bonnie Ray      F   9th   Huang High School   
5               5    Bryan Miranda      M   9th   Huang High School   
6               6    Sheena Carter      F  11th   Huang High School   
8               8     Michael Roth      M  10th   Huang High School   
9               9   Matthew Greene      M  10th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
6                 82          80          0  District  2917  1910635  
8                 95          87          0  District  2917  1910635  
9                 96          84          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[25528 rows x 11 columns]

In [60]:
# number of students who passed both math and reading.
per_passing_math_reading = per_passing_math_reading.groupby(["school_name"]).count()["student_name"]
per_passing_math_reading

school_name
Bailey High School       2719
Cabrera High School      1697
Figueroa High School     1569
Ford High School         1487
Griffin High School      1330
Hernandez High School    2481
Holden High School        381
Huang High School        1561
Johnson High School      2549
Pena High School          871
Rodriguez High School    2119
Shelton High School      1583
Thomas High School       1487
Wilson High School       2068
Wright High School       1626
Name: student_name, dtype: int64

In [61]:
# overall passing percentage.
per_overall_passing_percentage = per_passing_math_reading / per_school_counts * 100
per_overall_passing_percentage

school_name
Bailey High School       54.642283
Cabrera High School      91.334769
Figueroa High School     53.204476
Ford High School         54.289887
Griffin High School      90.599455
Hernandez High School    53.527508
Holden High School       89.227166
Huang High School        53.513884
Johnson High School      53.539172
Pena High School         90.540541
Rodriguez High School    52.988247
Shelton High School      89.892107
Thomas High School       90.948012
Wilson High School       90.582567
Wright High School       90.333333
dtype: float64

## Putting it all together into a single DF

In [62]:
per_school_summary_df = pd.DataFrame({
             "School Type": per_school_types,
             "Total Students": per_school_counts,
             "Total School Budget": per_school_budget,
             "Per Student Budget": per_school_capita,
             "Average Math Score": per_school_math,
           "Average Reading Score": per_school_reading,
           "% Passing Math": per_school_passing_math,
           "% Passing Reading": per_school_passing_reading,
           "% Overall Passing": per_overall_passing_percentage})
per_school_summary_df.head()

School Type  Total Students  Total School Budget  \
school_name                                                             
Bailey High School      District            4976              3124928   
Cabrera High School      Charter            1858              1081356   
Figueroa High School    District            2949              1884411   
Ford High School        District            2739              1763916   
Griffin High School      Charter            1468               917500   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Bailey High School                 628.0           77.048432   
Cabrera High School                582.0           83.061895   
Figueroa High School               639.0           76.711767   
Ford High School                   644.0           77.102592   
Griffin High School                625.0           83.351499   

                      Average Reading Score  % Passing Math  \
school_name                                                   
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
school_name                                                 
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

### Format columns

In [63]:
per_school_summary_df["Total School Budget"] = per_school_summary_df["Total School Budget"].map("${:,.2f}".format)

per_school_summary_df["Per Student Budget"] = per_school_summary_df["Per Student Budget"].map("${:,.2f}".format)


per_school_summary_df.head()

School Type  Total Students Total School Budget  \
school_name                                                            
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
school_name                                                   
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
school_name                                                   
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
school_name                                                 
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

### Reorder columns

In [64]:
new_column_order = ["School Type", "Total Students", "Total School Budget", "Per Student Budget", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]

per_school_summary_df = per_school_summary_df[new_column_order]

## Completed Summary

In [65]:
per_school_summary_df.head()

School Type  Total Students Total School Budget  \
school_name                                                            
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
school_name                                                   
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
school_name                                                   
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  
school_name                                                 
Bailey High School            81.933280          54.642283  
Cabrera High School           97.039828          91.334769  
Figueroa High School          80.739234          53.204476  
Ford High School              79.299014          54.289887  
Griffin High School           97.138965          90.599455

## Top five schools

In [66]:
top_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=False)

top_schools.head()

School Type  Total Students Total School Budget  \
school_name                                                           
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   
Pena High School        Charter             962         $585,858.00   

                    Per Student Budget  Average Math Score  \
school_name                                                  
Cabrera High School            $582.00           83.061895   
Thomas High School             $638.00           83.418349   
Griffin High School            $625.00           83.351499   
Wilson High School             $578.00           83.274201   
Pena High School               $609.00           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading  \
school_name                                                                     
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   
Pena High School                 84.044699       94.594595          95.945946   

                     % Overall Passing  
school_name                             
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

## Bottom 5 schools

In [67]:
# Sort and show top five schools.
bottom_schools = per_school_summary_df.sort_values(["% Overall Passing"], ascending=True)

bottom_schools.head()

School Type  Total Students Total School Budget  \
school_name                                                             
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Hernandez High School    District            4635       $3,022,020.00   
Johnson High School      District            4761       $3,094,650.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Rodriguez High School            $637.00           76.842711   
Figueroa High School             $639.00           76.711767   
Huang High School                $655.00           76.629414   
Hernandez High School            $652.00           77.289752   
Johnson High School              $650.00           77.072464   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Rodriguez High School          80.220055          52.988247  
Figueroa High School           80.739234          53.204476  
Huang High School              81.316421          53.513884  
Hernandez High School          80.862999          53.527508  
Johnson High School            81.222432          53.539172

### OOP Classes example

In [68]:
class Cat:
    def __init__(self, name):
        self.name = name

In [69]:
first_cat = Cat('Felix')
print(first_cat.name)

Felix


In [70]:
second_cat = Cat('Garfield')
print(second_cat.name)

Garfield


In [71]:
class Dog:
    def __init__(self, name, color, sound):
        self.name = name
        self.color = color
        self.sound = sound
    
    def bark(self):
        return self.sound + ' ' + self.sound

In [72]:
first_dog = Dog('Bella', 'brown', 'ruff!')
print(first_dog.name)
print(first_dog.color)
first_dog.bark()

Bella
brown


'ruff! ruff!'

In [73]:
second_dog = Dog('Lady', 'blonde', 'arf!')
second_dog.bark()

'arf! arf!'

### 4.10.1 Create Grade-Level DataFrames

Create two new DataFrames: one that will display the average math scores for each grade, and the one that will display the average reading scores for each grade

In [74]:
school_data_complete_df

Student ID     student_name gender grade         school_name  \
0               0     Paul Bradley      M   9th   Huang High School   
1               1     Victor Smith      M  12th   Huang High School   
2               2  Kevin Rodriguez      M  12th   Huang High School   
3               3    Richard Scott      M  12th   Huang High School   
4               4       Bonnie Ray      F   9th   Huang High School   
...           ...              ...    ...   ...                 ...   
39165       39165     Donna Howard      F  12th  Thomas High School   
39166       39166        Dawn Bell      F  10th  Thomas High School   
39167       39167   Rebecca Tanner      F   9th  Thomas High School   
39168       39168     Desiree Kidd      F  10th  Thomas High School   
39169       39169  Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
3                 67          58          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[39170 rows x 11 columns]

In [75]:
ninth_graders = school_data_complete_df[(school_data_complete_df["grade"]=="9th")]
ninth_graders

Student ID      student_name gender grade         school_name  \
0               0      Paul Bradley      M   9th   Huang High School   
4               4        Bonnie Ray      F   9th   Huang High School   
5               5     Bryan Miranda      M   9th   Huang High School   
12             12   Brittney Walker      F   9th   Huang High School   
13             13      William Long      M   9th   Huang High School   
...           ...               ...    ...   ...                 ...   
39152       39152        Lori Moore      F   9th  Thomas High School   
39153       39153   William Hubbard      M   9th  Thomas High School   
39157       39157  Kristen Gonzalez      F   9th  Thomas High School   
39164       39164    Joseph Anthony      M   9th  Thomas High School   
39167       39167    Rebecca Tanner      F   9th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
5                 94          94          0  District  2917  1910635  
12                64          79          0  District  2917  1910635  
13                71          79          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39152             98          84         14   Charter  1635  1043130  
39153             80          75         14   Charter  1635  1043130  
39157             79          94         14   Charter  1635  1043130  
39164             97          76         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  

[11408 rows x 11 columns]

In [76]:
tenth_graders = school_data_complete_df[(school_data_complete_df["grade"]=="10th")]
tenth_graders

Student ID      student_name gender grade         school_name  \
8               8      Michael Roth      M  10th   Huang High School   
9               9    Matthew Greene      M  10th   Huang High School   
10             10  Andrew Alexander      M  10th   Huang High School   
11             11     Daniel Cooper      M  10th   Huang High School   
14             14      Tammy Hebert      F  10th   Huang High School   
...           ...               ...    ...   ...                 ...   
39155       39155       John Brooks      M  10th  Thomas High School   
39158       39158     Kari Holloway      F  10th  Thomas High School   
39161       39161       April Reyes      F  10th  Thomas High School   
39166       39166         Dawn Bell      F  10th  Thomas High School   
39168       39168      Desiree Kidd      F  10th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
8                 95          87          0  District  2917  1910635  
9                 96          84          0  District  2917  1910635  
10                90          70          0  District  2917  1910635  
11                78          77          0  District  2917  1910635  
14                85          67          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39155             92          98         14   Charter  1635  1043130  
39158             87          90         14   Charter  1635  1043130  
39161             70          84         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  

[10168 rows x 11 columns]

In [77]:
eleventh_graders = school_data_complete_df[(school_data_complete_df["grade"]=="11th")]
eleventh_graders

Student ID         student_name gender grade         school_name  \
6               6        Sheena Carter      F  11th   Huang High School   
15             15        Jordan Carson      M  11th   Huang High School   
22             22     Kimberly Stewart      F  11th   Huang High School   
24             24     Chelsea Griffith      F  11th   Huang High School   
28             28          Kelly James      F  11th   Huang High School   
...           ...                  ...    ...   ...                 ...   
39156       39156  Stephanie Contreras      F  11th  Thomas High School   
39159       39159     Kimberly Cabrera      F  11th  Thomas High School   
39160       39160         Katie Weaver      F  11th  Thomas High School   
39163       39163           John Reese      M  11th  Thomas High School   
39169       39169      Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
6                 82          80          0  District  2917  1910635  
15                94          88          0  District  2917  1910635  
22                99          84          0  District  2917  1910635  
24                85          73          0  District  2917  1910635  
28                73          55          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39156             79          95         14   Charter  1635  1043130  
39159             85          72         14   Charter  1635  1043130  
39160             89          86         14   Charter  1635  1043130  
39163             90          75         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[9695 rows x 11 columns]

In [78]:
twelfth_graders = school_data_complete_df[(school_data_complete_df["grade"]=="12th")]
twelfth_graders

Student ID      student_name gender grade         school_name  \
1               1      Victor Smith      M  12th   Huang High School   
2               2   Kevin Rodriguez      M  12th   Huang High School   
3               3     Richard Scott      M  12th   Huang High School   
7               7      Nicole Baker      F  12th   Huang High School   
29             29      Nicole Brown      F  12th   Huang High School   
...           ...               ...    ...   ...                 ...   
39137       39137   Victoria Bailey      F  12th  Thomas High School   
39144       39144  Cristian Webster      M  12th  Thomas High School   
39154       39154   Bradley Johnson      M  12th  Thomas High School   
39162       39162       Derek Weeks      M  12th  Thomas High School   
39165       39165      Donna Howard      F  12th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
3                 67          58          0  District  2917  1910635  
7                 96          69          0  District  2917  1910635  
29                90          88          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39137             75          70         14   Charter  1635  1043130  
39144             77          87         14   Charter  1635  1043130  
39154             91          71         14   Charter  1635  1043130  
39162             94          77         14   Charter  1635  1043130  
39165             99          90         14   Charter  1635  1043130  

[7899 rows x 11 columns]

### @ 4.10.2

In [79]:
ninth_grade_math_scores = ninth_graders.groupby(["school_name"]).mean()["math_score"]
tenth_grade_math_scores = tenth_graders.groupby(["school_name"]).mean()["math_score"]
eleventh_grade_math_scores = eleventh_graders.groupby(["school_name"]).mean()["math_score"]
twelfth_grade_math_scores = twelfth_graders.groupby(["school_name"]).mean()["math_score"]

ninth_grade_reading_scores = ninth_graders.groupby(["school_name"]).mean()["reading_score"]
tenth_grade_reading_scores = tenth_graders.groupby(["school_name"]).mean()["reading_score"]
eleventh_grade_reading_scores = eleventh_graders.groupby(["school_name"]).mean()["reading_score"]
twelfth_grade_reading_scores = twelfth_graders.groupby(["school_name"]).mean()["reading_score"]

### Put the scores together in to DataFrames

In [80]:
math_scores_by_grade = pd.DataFrame({
               "9th": ninth_grade_math_scores,
               "10th": tenth_grade_math_scores,
               "11th": eleventh_grade_math_scores,
               "12th": twelfth_grade_math_scores})

reading_scores_by_grade = pd.DataFrame({
               "9th": ninth_grade_reading_scores,
               "10th": tenth_grade_reading_scores,
               "11th": eleventh_grade_reading_scores,
               "12th": twelfth_grade_reading_scores})

### Format scores to 2 decimal

In [81]:
math_scores_by_grade["9th"] = math_scores_by_grade["9th"].map("{:.1f}".format)
math_scores_by_grade["10th"] = math_scores_by_grade["10th"].map("{:.1f}".format)
math_scores_by_grade["11th"] = math_scores_by_grade["11th"].map("{:.1f}".format)
math_scores_by_grade["12th"] = math_scores_by_grade["12th"].map("{:.1f}".format)

reading_scores_by_grade["9th"] = reading_scores_by_grade["9th"].map("{:.1f}".format)
reading_scores_by_grade["10th"] = reading_scores_by_grade["10th"].map("{:.1f}".format)
reading_scores_by_grade["11th"] = reading_scores_by_grade["11th"].map("{:.1f}".format)
reading_scores_by_grade["12th"] = reading_scores_by_grade["12th"].map("{:.1f}".format)

### Order columns

In [82]:
math_scores_by_grade = math_scores_by_grade[["9th", "10th", "11th", "12th"]]
reading_scores_by_grade = math_scores_by_grade[["9th", "10th", "11th", "12th"]]

### Remove the index name

In [83]:
math_scores_by_grade.index.name = None
reading_scores_by_grade.index.name = None

In [84]:
math_scores_by_grade.head()

9th  10th  11th  12th
Bailey High School    77.1  77.0  77.5  76.5
Cabrera High School   83.1  83.2  82.8  83.3
Figueroa High School  76.4  76.5  76.9  77.2
Ford High School      77.4  77.7  76.9  76.2
Griffin High School   82.0  84.2  83.8  83.4

In [85]:
reading_scores_by_grade.head()

9th  10th  11th  12th
Bailey High School    77.1  77.0  77.5  76.5
Cabrera High School   83.1  83.2  82.8  83.3
Figueroa High School  76.4  76.5  76.9  77.2
Ford High School      77.4  77.7  76.9  76.2
Griffin High School   82.0  84.2  83.8  83.4

## 4.11.1 Spending Ranges per bin

In [86]:
per_school_capita

school_name
Huang High School        655.0
Figueroa High School     639.0
Shelton High School      600.0
Hernandez High School    652.0
Griffin High School      625.0
Wilson High School       578.0
Cabrera High School      582.0
Bailey High School       628.0
Holden High School       581.0
Pena High School         609.0
Wright High School       583.0
Rodriguez High School    637.0
Johnson High School      650.0
Ford High School         644.0
Thomas High School       638.0
dtype: float64

In [87]:
per_school_capita.describe()

count     15.000000
mean     620.066667
std       28.544368
min      578.000000
25%      591.500000
50%      628.000000
75%      641.500000
max      655.000000
dtype: float64

### Why are these ranges what they are given the numbers above

Seems arbitrary. Understand we don't want 578 as first bin because there is only one. With the std being 28, why does 578 become 585 and 641 become 675? Or the aim is to attempt to group evenly as possible?

585 - 578 = 7

615 - 591 = 24

645 - 628 = 17

675 - 641 = 34

In [88]:
spending_bins = [0, 585, 630, 645, 675]

The cut() function places each value of the per_school_capita Series into one of four bins. The bins are created by a list of five numbers given as the spending_bins that define the edges.

In [89]:
pd.cut(per_school_capita, spending_bins)

school_name
Huang High School        (645, 675]
Figueroa High School     (630, 645]
Shelton High School      (585, 630]
Hernandez High School    (645, 675]
Griffin High School      (585, 630]
Wilson High School         (0, 585]
Cabrera High School        (0, 585]
Bailey High School       (585, 630]
Holden High School         (0, 585]
Pena High School         (585, 630]
Wright High School         (0, 585]
Rodriguez High School    (630, 645]
Johnson High School      (645, 675]
Ford High School         (630, 645]
Thomas High School       (630, 645]
dtype: category
Categories (4, interval[int64]): [(0, 585] < (585, 630] < (630, 645] < (645, 675]]

In [90]:
per_school_capita.groupby(pd.cut(per_school_capita, spending_bins)).count()

(0, 585]      4
(585, 630]    4
(630, 645]    4
(645, 675]    3
dtype: int64

## Rename bins

In [91]:
group_names = ["<$584", "$585-629", "$630-644", "$645-675"]

In [92]:
# Categorize spending based on the bins.
per_school_summary_df["Spending Ranges (Per Student)"] = pd.cut(per_school_capita, spending_bins, labels=group_names)

per_school_summary_df

School Type  Total Students Total School Budget  \
school_name                                                             
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Bailey High School               $628.00           77.048432   
Cabrera High School              $582.00           83.061895   
Figueroa High School             $639.00           76.711767   
Ford High School                 $644.00           77.102592   
Griffin High School              $625.00           83.351499   
Hernandez High School            $652.00           77.289752   
Holden High School               $581.00           83.803279   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Pena High School                 $609.00           83.839917   
Rodriguez High School            $637.00           76.842711   
Shelton High School              $600.00           83.359455   
Thomas High School               $638.00           83.418349   
Wilson High School               $578.00           83.274201   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing  \
school_name                                                   
Bailey High School             81.933280          54.642283   
Cabrera High School            97.039828          91.334769   
Figueroa High School           80.739234          53.204476   
Ford High School               79.299014          54.289887   
Griffin High School            97.138965          90.599455   
Hernandez High School          80.862999          53.527508   
Holden High School             96.252927          89.227166   
Huang High School              81.3

In [93]:
# Calculate averages for the desired columns.
spending_math_scores = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Math Score"]

spending_reading_scores = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Reading Score"]

spending_passing_math = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Math"]

spending_passing_reading = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Reading"]

overall_passing_spending = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Overall Passing"]

In [94]:
overall_passing_spending

Spending Ranges (Per Student)
<$584       90.369459
$585-629    81.418596
$630-644    62.857656
$645-675    53.526855
Name: % Overall Passing, dtype: float64

In [95]:
per_school_summary_df.head()

School Type  Total Students Total School Budget  \
school_name                                                            
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
school_name                                                   
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
school_name                                                   
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  \
school_name                                                  
Bailey High School            81.933280          54.642283   
Cabrera High School           97.039828          91.334769   
Figueroa High School          80.739234          53.204476   
Ford High School              79.299014          54.289887   
Griffin High School           97.138965          90.599455   

                     Spending Ranges (Per Student)  
school_name                                         
Bailey High School                        $585-629  
Cabrera High School                          <$584  
Figueroa High School                      $630-644  
Ford High School                          $630-644  
Griffin High School                       $585-629

In [96]:
overall_passing_count = per_school_summary_df.groupby(["Spending Ranges (Per Student)"]).count()['School Type']
overall_passing_count

Spending Ranges (Per Student)
<$584       4
$585-629    4
$630-644    4
$645-675    3
Name: School Type, dtype: int64

In [97]:
spending_summary_df = pd.DataFrame({
          "Average Math Score" : spending_math_scores,
          "Average Reading Score": spending_reading_scores,
          "% Passing Math": spending_passing_math,
          "% Passing Reading": spending_passing_reading,
          "% Overall Passing": overall_passing_spending})

spending_summary_df

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$584                                   83.455399              83.933814   
$585-629                                81.899826              83.155286   
$630-644                                78.518855              81.624473   
$645-675                                76.997210              81.027843   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$584                               93.460096          96.610877   
$585-629                            87.133538          92.718205   
$630-644                            73.484209          84.391793   
$645-675                            66.164813          81.133951   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$584                                  90.369459  
$585-629                               81.418596  
$630-644                               62.857656  
$645-675                               53.526855

## 4.11.4 Formatting DF

In [98]:
spending_summary_df["Average Math Score"] = spending_summary_df["Average Math Score"].map("{:.1f}".format)
spending_summary_df["Average Reading Score"] = spending_summary_df["Average Reading Score"].map("{:.1f}".format)
spending_summary_df["% Passing Math"] = spending_summary_df["% Passing Math"].map("{:.0f}".format)
spending_summary_df["% Passing Reading"] = spending_summary_df["% Passing Reading"].map("{:.0f}".format)
spending_summary_df["% Overall Passing"] = spending_summary_df["% Overall Passing"].map("{:.0f}".format)

spending_summary_df

Average Math Score Average Reading Score  \
Spending Ranges (Per Student)                                            
<$584                                       83.5                  83.9   
$585-629                                    81.9                  83.2   
$630-644                                    78.5                  81.6   
$645-675                                    77.0                  81.0   

                              % Passing Math % Passing Reading  \
Spending Ranges (Per Student)                                    
<$584                                     93                97   
$585-629                                  87                93   
$630-644                                  73                84   
$645-675                                  66                81   

                              % Overall Passing  
Spending Ranges (Per Student)                    
<$584                                        90  
$585-629                                     81  
$630-644                                     63  
$645-675                                     54

## 4.12.1 Create Bins for School Size

create three bins: small, medium, and large.

In [99]:
per_school_counts = school_data_complete_df["school_name"].value_counts()
per_school_counts

Bailey High School       4976
Johnson High School      4761
Hernandez High School    4635
Rodriguez High School    3999
Figueroa High School     2949
Huang High School        2917
Ford High School         2739
Wilson High School       2283
Cabrera High School      1858
Wright High School       1800
Shelton High School      1761
Thomas High School       1635
Griffin High School      1468
Pena High School          962
Holden High School        427
Name: school_name, dtype: int64

Human decision making:

"Small" schools have fewer than 1,000 students

"Medium" schools have 1,000–1,999 students

"Large" schools have 2,000–5,000 students

### Create the series to cut() with

Use these 2 series to cut() the 'per_school_summary_df'

In [100]:
size_bins = [0, 1000, 2000, 5000]
group_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

### Add new column "School Size"

- cut into per_school_summary_df
- (param1) add new column "School Size"
- (param2) range
- (param3) labels

In [101]:
per_school_summary_df["School Size"] = pd.cut(per_school_summary_df["Total Students"], size_bins, labels=group_names)
per_school_summary_df.head()

School Type  Total Students Total School Budget  \
school_name                                                            
Bailey High School      District            4976       $3,124,928.00   
Cabrera High School      Charter            1858       $1,081,356.00   
Figueroa High School    District            2949       $1,884,411.00   
Ford High School        District            2739       $1,763,916.00   
Griffin High School      Charter            1468         $917,500.00   

                     Per Student Budget  Average Math Score  \
school_name                                                   
Bailey High School              $628.00           77.048432   
Cabrera High School             $582.00           83.061895   
Figueroa High School            $639.00           76.711767   
Ford High School                $644.00           77.102592   
Griffin High School             $625.00           83.351499   

                      Average Reading Score  % Passing Math  \
school_name                                                   
Bailey High School                81.033963       66.680064   
Cabrera High School               83.975780       94.133477   
Figueroa High School              81.158020       65.988471   
Ford High School                  80.746258       68.309602   
Griffin High School               83.816757       93.392371   

                      % Passing Reading  % Overall Passing  \
school_name                                                  
Bailey High School            81.933280          54.642283   
Cabrera High School           97.039828          91.334769   
Figueroa High School          80.739234          53.204476   
Ford High School              79.299014          54.289887   
Griffin High School           97.138965          90.599455   

                     Spending Ranges (Per Student)         School Size  
school_name                                                             
Bailey High School                        $585-629   Large (2000-5000)  
Cabrera High School                          <$584  Medium (1000-2000)  
Figueroa High School                      $630-644   Large (2000-5000)  
Ford High School                          $630-644   Large (2000-5000)  
Griffin High School                       $585-629  Medium (1000-2000)

## 4.12.3 Group by School Size

In [102]:
# calculate averages 
size_math_scores = per_school_summary_df.groupby(["School Size"]).mean()["Average Math Score"]
size_reading_scores = per_school_summary_df.groupby(["School Size"]).mean()["Average Reading Score"]

size_passing_math = per_school_summary_df.groupby(["School Size"]).mean()["% Passing Math"]
size_passing_reading = per_school_summary_df.groupby(["School Size"]).mean()["% Passing Reading"]

size_overall_passing = per_school_summary_df.groupby(["School Size"]).mean()["% Overall Passing"]

### Put them together 

In [103]:
size_summary_df = pd.DataFrame({
          "Average Math Score" : size_math_scores,
          "Average Reading Score": size_reading_scores,
          "% Passing Math": size_passing_math,
          "% Passing Reading": size_passing_reading,
          "% Overall Passing": size_overall_passing})

size_summary_df

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)               96.099437          89.883853  
Medium (1000-2000)          96.790680          90.621535  
Large (2000-5000)           82.766634          58.286003

In [104]:
size_summary_df['Average Math Score'] = size_summary_df['Average Math Score'].map("{:.1f}".format)
size_summary_df['Average Reading Score'] = size_summary_df['Average Reading Score'].map("{:.1f}".format)
size_summary_df['% Passing Math'] = size_summary_df['% Passing Math'].map("{:.0f}".format)
size_summary_df['% Passing Reading'] = size_summary_df['% Passing Reading'].map("{:.0f}".format)
size_summary_df['% Overall Passing'] = size_summary_df['% Overall Passing'].map("{:.0f}".format)

In [105]:
size_summary_df

Average Math Score Average Reading Score % Passing Math  \
School Size                                                                  
Small (<1000)                    83.8                  83.9             94   
Medium (1000-2000)               83.4                  83.9             94   
Large (2000-5000)                77.7                  81.3             70   

                   % Passing Reading % Overall Passing  
School Size                                             
Small (<1000)                     96                90  
Medium (1000-2000)                97                91  
Large (2000-5000)                 83                58